experimental/Matrix/matrix.jl

```diff
@@ -0,0 +1,112 @@
+module MatrixGroups
+
+using GAP
+using Oscar
+
+export _wrap_for_gap
+  
+################################################################################
+# Initilize GAP function, i.e. GAP reads the file matrix.g
+#
+function __init__()
+    GAP.Globals.Reread(GAP.GapObj(joinpath(Oscar.oscardir, "experimental", "Matrix", "matrix.g")))
+end
+
+################################################################################
+# Computes the JuliaMatrixRep of a MatrixElem in GAP
+#
+#    _wrap_for_gap(m::MatrixElem)
+#
+# Input: m::MatrixElem
+# Output: Returns the JuliaMatrixRep of m in GAP
+#
+# Example
+# ```jldoctest
+# julia> m = matrix(ZZ, [0 1 ; -1 0]);
+# julia> _wrap_for_gap(m)
+# GAP: <matrix object of dimensions 2x2 over Integer Ring>
+# ```
+# """
+function _wrap_for_gap(m::MatrixElem)
+    return GAP.Globals.MakeJuliaMatrixRep(m)
+end
+
+
+################################################################################
+# Computes a GAP Group where the generators are the JuliaMatrixRep of matrices.
+# Moreover, a nice morphism from G into a GAP matrix group over a finite field G2 is constructed
+# such that calculations in G can be handled by G2 automatically.
+#
+#     MatrixGroup(matrices::Vector{<:MatrixElem{T}}) where T <: Union{fmpz, fmpq, nf_elem}
+#
+# Input: matrices::Vector{<:MatrixElem{T}} where T <: Union{fmpz, fmpq, nf_elem}
+# Output: GAP group generated by the JuliaMatrixReps of matrices
+#
+# Example
+# ```jldoctest
+# julia> m1 = matrix(QQ, [0 1 ; -1 0]);
+# julia> m2 = matrix(QQ, [ -1 0; 0 1]);
+# julia> MatrixGroup([m1,m2])
+# GAP: <group with 2 generators>
+# ```
+# """
+function MatrixGroup(matrices::Vector{<:MatrixElem{T}}) where T <: Union{fmpz, fmpq, nf_elem}
+       @assert !isempty(matrices)
+    
+       # One should probably check whether all matrices are n by n (and invertible
+       # and such ...)
+
+       K = base_ring(matrices[1])
+       if K isa FlintIntegerRing
+          K = QQ
+       end
+       for mat in matrices
+            if !(K == mat.base_ring)
+                error("Matrices are not from the same base ring")
+            end
+       end
+       n = nrows(matrices[1])
+
+       Fq, matrices_Fq, OtoFq = Oscar.good_reduction(matrices, 2)
+
+       G = Oscar.MatrixGroup(n, Fq, matrices_Fq)
+       N = order(G)
```

Comment:
I don't think we should call `MatrixGroup` here. Rather, something like this pseudocode
```
  gap_matrices_Fq = [map_from_julia_to_gap(m) for m in matrices_Fq]
  G2 = GAP.Globals.Group(gap_matrices_Fq)
  N = fmpz(GAPWrap.Order(G2))
```
This way we avoid creating that wrapper G we don't need later on anyway